### PR TITLE
crypto/provider_core.c: correct definition of lock

### LINKED ISSUE #8476: provider_core.c not compiling if HAVE_ATOMICS is not defined
Build Failure is:
gcc  -I. -Icrypto/include -Iinclude -I.. -I../crypto/include -I../include  -fPIC -pthread -m32 -Wa,--noexecstack -Wall -O3 -fomit-frame-pointer -DOPENSSL_USE_NODELETE -DL_ENDIAN -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_BN_ASM_PART_WORDS -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DRC4_ASM -DMD5_ASM -DRMD160_ASM -DAES_ASM -DVPAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM -DPOLY1305_ASM -DOPENSSLDIR="\"/home/nse/workspace/openssl-linux_x86_gcc/install/ssl\"" -DENGINESDIR="\"/home/nse/workspace/openssl-linux_x86_gcc/install/lib/engines-3\"" -DMODULESDIR="\"/home/nse/workspace/openssl-linux_x86_gcc/install/lib/ossl-modules\"" -DNDEBUG  -MMD -MF crypto/libcrypto-lib-provider_core.d.tmp -MT crypto/libcrypto-lib-provider_core.o -c -o crypto/libcrypto-lib-provider_core.o ../crypto/provider_core.c
../crypto/provider_core.c:32: error: variable or field 'refcnt_lock' declared void
../crypto/provider_core.c: In function 'ossl_provider_upref':
../crypto/provider_core.c:127: warning: passing argument 4 of 'CRYPTO_atomic_add' makes pointer from integer without a cast
../crypto/provider_core.c: In function 'ossl_provider_free':
../crypto/provider_core.c:211: error: 'provider_lock' undeclared (first use in this function)
../crypto/provider_core.c:211: error: (Each undeclared identifier is reported only once
../crypto/provider_core.c:211: error: for each function it appears in.)

An example build config that is failing (there are many)
--------------------------------------------------------
/workspace/openssl-linux_x86_gcc/Configure --prefix=/workspace/openssl-linux_x86_gcc/install shared linux-x86

Command line (with current working directory = .):

    /usr/local/bin/perl ../Configure --prefix=/home/nse/workspace/openssl-linux_x86_gcc/install shared linux-x86

Perl information:

    /usr/local/bin/perl
    5.24.0 for x86_64-linux

Enabled features:

    aria
    asm
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    camellia
    capieng
    cast
    chacha
    cmac
    cms
    comp
    crmf
    ct
    deprecated
    des
    dgram
    dh
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    engine
    err
    filenames
    gost
    idea
    makedepend
    md4
    mdc2
    multiblock
    nextprotoneg
    pinshared
    ocb
    ocsp
    padlockeng
    pic
    poly1305
    posix-io
    psk
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    seed
    shared
    siphash
    siv
    sm2
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    static-engine
    stdio
    tests
    threads
    tls
    trace
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    ktls                    [default]        OPENSSL_NO_KTLS
    afalgeng                [too-old-kernel] 
    asan                    [default]        OPENSSL_NO_ASAN
    buildtest-c++           [default]        
    crypto-mdebug           [default]        OPENSSL_NO_CRYPTO_MDEBUG
    crypto-mdebug-backtrace [default]        OPENSSL_NO_CRYPTO_MDEBUG_BACKTRACE
    devcryptoeng            [default]        OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128     [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    egd                     [default]        OPENSSL_NO_EGD
    external-tests          [default]        OPENSSL_NO_EXTERNAL_TESTS
    fuzz-libfuzzer          [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    fuzz-afl                [default]        OPENSSL_NO_FUZZ_AFL
    heartbeats              [default]        OPENSSL_NO_HEARTBEATS
    md2                     [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                    [default]        OPENSSL_NO_MSAN
    rc5                     [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                    [default]        OPENSSL_NO_SCTP
    ssl-trace               [default]        OPENSSL_NO_SSL_TRACE
    ubsan                   [default]        OPENSSL_NO_UBSAN
    unit-test               [default]        OPENSSL_NO_UNIT_TEST
    weak-ssl-ciphers        [default]        OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                    [default]        
    zlib-dynamic            [default]        
    ssl3                    [default]        OPENSSL_NO_SSL3
    ssl3-method             [default]        OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "r",
    CC => "gcc",
    CFLAGS => "-Wall -O3 -fomit-frame-pointer",
    CXX => "g++",
    CXXFLAGS => "-Wall -O3",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    aes_asm_src => "aes-586.s vpaes-x86.s aesni-x86.s",
    aes_obj => "aes-586.o vpaes-x86.o aesni-x86.o",
    apps_aux_src => "",
    apps_init_src => "",
    apps_obj => "",
    bf_asm_src => "bf-586.s",
    bf_obj => "bf-586.o",
    bn_asm_src => "bn-586.s co-586.s x86-mont.s x86-gf2m.s",
    bn_obj => "bn-586.o co-586.o x86-mont.o x86-gf2m.o",
    bn_ops => "BN_LLONG",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cast_asm_src => "c_enc.c",
    cast_obj => "c_enc.o",
    cflags => "-pthread -m32",
    chacha_asm_src => "chacha-x86.s",
    chacha_obj => "chacha-x86.o",
    cmll_asm_src => "cmll-x86.s",
    cmll_obj => "cmll-x86.o",
    cppflags => "",
    cpuid_asm_src => "x86cpuid.s",
    cpuid_obj => "x86cpuid.o",
    cxxflags => "-std=c++11 -pthread -m32",
    defines => [  ],
    des_asm_src => "des-586.s crypt586.s",
    des_obj => "des-586.o crypt586.o",
    disable => [  ],
    dso_scheme => "dlfcn",
    ec_asm_src => "ecp_nistz256.c ecp_nistz256-x86.s",
    ec_obj => "ecp_nistz256.o ecp_nistz256-x86.o",
    enable => [ "afalgeng" ],
    ex_libs => "-ldl -pthread",
    includes => [  ],
    keccak1600_asm_src => "keccak1600.c",
    keccak1600_obj => "keccak1600.o",
    lflags => "",
    lib_cflags => "",
    lib_cppflags => "-DOPENSSL_USE_NODELETE -DL_ENDIAN",
    lib_defines => [  ],
    md5_asm_src => "md5-586.s",
    md5_obj => "md5-586.o",
    modes_asm_src => "ghash-x86.s",
    modes_obj => "ghash-x86.o",
    module_cflags => "-fPIC",
    module_cxxflags => "",
    module_ldflags => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    padlock_asm_src => "e_padlock-x86.s",
    padlock_obj => "e_padlock-x86.o",
    perl_platform => "Unix",
    perlasm_scheme => "elf",
    poly1305_asm_src => "poly1305-x86.s",
    poly1305_obj => "poly1305-x86.o",
    rc4_asm_src => "rc4-586.s",
    rc4_obj => "rc4-586.o",
    rc5_asm_src => "rc5-586.s",
    rc5_obj => "rc5-586.o",
    rmd160_asm_src => "rmd-586.s",
    rmd160_obj => "rmd-586.o",
    sha1_asm_src => "sha1-586.s sha256-586.s sha512-586.s",
    sha1_obj => "sha1-586.o sha256-586.o sha512-586.o",
    shared_cflag => "-fPIC",
    shared_defflag => "-Wl,--version-script=",
    shared_defines => [  ],
    shared_ldflag => "-Wl,-znodelete -shared -Wl,-Bsymbolic",
    shared_rcflag => "",
    shared_sonameflag => "-Wl,-soname=",
    shared_target => "linux-shared",
    thread_defines => [  ],
    thread_scheme => "pthreads",
    unistd => "<unistd.h>",
    uplink_aux_src => "",
    uplink_obj => "",
    wp_asm_src => "wp_block.c wp-mmx.s",
    wp_obj => "wp_block.o wp-mmx.o",

Recorded environment:

    AR = 
    ARFLAGS = 
    AS = 
    ASFLAGS = 
    BUILDFILE = 
    CC = 
    CFLAGS = 
    CPP = 
    CPPDEFINES = 
    CPPFLAGS = 
    CPPINCLUDES = 
    CROSS_COMPILE = 
    CXX = 
    CXXFLAGS = 
    HASHBANGPERL = 
    LD = 
    LDFLAGS = 
    LDLIBS = 
    MT = 
    MTFLAGS = 
    OPENSSL_LOCAL_CONFIG_DIR = 
    PERL = 
    RANLIB = 
    RC = 
    RCFLAGS = 
    RM = 
    WINDRES = 
    __CNF_CFLAGS = 
    __CNF_CPPDEFINES = 
    __CNF_CPPFLAGS = 
    __CNF_CPPINCLUDES = 
    __CNF_CXXFLAGS = 
    __CNF_LDFLAGS = 
    __CNF_LDLIBS = 

Makevars:

    AR              = ar
    ARFLAGS         = r
    CC              = gcc
    CFLAGS          = -Wall -O3 -fomit-frame-pointer
    CPPDEFINES      = 
    CPPFLAGS        = 
    CPPINCLUDES     = 
    CXX             = g++
    CXXFLAGS        = -Wall -O3
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         = 
    LDLIBS          = 
    PERL            = /usr/local/bin/perl
    RANLIB          = ranlib
    RC              = windres

### PR DESCRIPTION
**Pull request by levitte**
Fixes #8476

### CODE REVIEW

**Review by paulidale — DISMISSED** [review]

File: crypto/provider_core.c [hunk 1]
```diff
@@ -208,7 +208,7 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
         int ref = 0;
 
 #ifndef HAVE_ATOMICS
-        CRYPTO_DOWN_REF(&prov->refcnt, &ref, provider_lock);
+        CRYPTO_DOWN_REF(&prov->refcnt, &ref, prov->refcnt_lock);
```
Comment (slontis) [inline crypto/provider_core.c]: refcnt_lock appears to be NULL?

Comment (levitte) [reply crypto/provider_core.c]: Argh...

The refcount header doesn't make it easy to build without atomics :confused: 
That needs to change.

Comment (levitte) [reply crypto/provider_core.c]: Fixed

Comment (mattcaswell) [reply crypto/provider_core.c]: I do wonder why we're using `HAVE_ATOMICS` here at all. We could just create the lock unconditionally. If HAVE_ATOMICS is defined then it won't be used by the `CRYPTO_DOWN_REF` implementation...but so what?

Comment (levitte) [reply crypto/provider_core.c]: 'til some code sanitizer notices and starts complaining about it

Comment (mattcaswell) [reply crypto/provider_core.c]: It's what we do *everywhere* else. This is the only file in the whole repo that actually *uses* the HAVE_ATOMICS define.

Comment (levitte) [reply crypto/provider_core.c]: Okie.  I've now ripped out most of those ifdefs.  I will *refuse* to remove those around `CRYPTO_THREAD_lock_new` and `CRYPTO_THREAD_lock_free`, though.  That would be stoopid waste of resources (and if we *are* wasteful elsewhere, I think guards should be *added*)

File: crypto/provider_core.c [hunk 2]
```diff
@@ -231,6 +224,9 @@ void ossl_provider_free(OSSL_PROVIDER *prov)
         if (ref == 0) {
             DSO_free(prov->module);
             OPENSSL_free(prov->name);
+#ifndef HAVE_ATOMICS
+            CRYPTO_THREAD_lock_free(prov->refcnt_lock);
+#endif
```
Comment (mattcaswell) [inline crypto/provider_core.c]: You could also make this unconditional

**Review by mattcaswell — APPROVED** [review]
Approved whether or not you make the suggested change

File: crypto/provider_core.c (hunk 2 above)
Comment (levitte) [reply crypto/provider_core.c]: I'll have a closer look at the HAVE_ATOMICS stuff.  We could do it better

**Comment by levitte** [comment]
Merged.

085bef9f7cee8506cc9d438d4fbc3663f46c5fe2 crypto/provider_core.c: correct definition and use of lock

**Comment by richsalz** [comment]
Should CRYPTO_LOCK be used in the same in all API's no matter what ifdef's are used?  Isn't that the point of the abstraction?

**Comment by levitte** [comment]
> Should CRYPTO_LOCK be used in the same in all API's no matter what ifdef's are used? Isn't that the point of the abstraction?

We ended up doing that here as well.  The only thing I still guard is the creation of the lock.  There's no point creating it if it won't be used...

**Comment by richsalz** [comment]
Oops, I misread the diff.  Yes, you're completely right and I wasted time.

File: crypto/provider_core.c (hunk 2 above)
Comment (paulidale) [reply crypto/provider_core.c]: I planned on producing a macro or two that do the necessary conditioning for us.
It shouldn't be hard to make this a lot easier.

**Pull request closed without merging**